### PR TITLE
fix: reduce UnboundMethod objects by memoizing initialize_parameters

### PR DESCRIPTION
**Pull request by rainerborene**
I have just discovered a simple optimization on `ViewComponent::Base` that can reduce `UnboundMethod` objects allocation. I have removed other objects from this memory allocation report comparison to keep things simple and clear. These are the results in one of our main Rails endpoints.

Before:
```
allocated memory by class
-------------------------
4080  UnboundMethod
```

After:
```
allocated memory by class
-------------------------
1280  UnboundMethod
```

If you want to test this patch yourself, just do this on your `ApplicationComponent`:

```ruby
class ApplicationComponent < ViewComponent::Base
  private
    def self.initialize_parameters
      @initialize_parameters ||= super
    end
end
```

### CODE REVIEW

**Comment by reeganviljoen** [comment]
@rainerborene can you please add your change to the changelog while waiting on a review

**Comment by Spone** [comment]
I'm a bit unsure if adding memorization here can have unwanted effects. Mind chiming in @camertron?

**Comment by camertron** [comment]
Cool, thanks for looking into this @rainerborene :)

The only issue I can think of would be if the `#initialize` method gets redefined via `class_eval` or something. It's possible in such a scenario that the memoized list of parameters would be out-of-date by the time `Base` tries to find the collection parameter. However I can't imagine that sort of thing is very common, so I think this change is ok 👍

**Comment by camertron** [comment]
I'm on leave right now @Spone, would you or another maintainer be able to look into the Vale CI failure and merge this?

**Comment by Spone** [comment]
Yes I will, thanks for your input!

File: lib/view_component/base.rb
Comment (camertron) [inline lib/view_component/base.rb:651]: Please set `@initialize_parameters = nil` in `.with_collection_parameter` above to cover the case where the collection parameter is changed after memoization.

**Review by camertron — CHANGES_REQUESTED** [review]

File: lib/view_component/base.rb
Comment (rainerborene) [reply lib/view_component/base.rb:651]: @camertron `initialize_parameters` is not inherited. I don't think there is a need to set it to `nil`. Even if `.with_collection_parameter` is called the `initialize` method was already defined in the first place. Can you plz give me an example to reproduce the issue you're talking about?

Comment (camertron) [reply lib/view_component/base.rb:651]: Consider this (admittedly contrived) example:

```ruby
module Primer
  class ProductComponent < Primer::Component
    def initialize(product: nil)
      @product = product
    end
  end
end

# ensure parameters are memoized
Primer::ProductComponent.send(:initialize_parameters)

Primer::ProductComponent.class_eval do
  with_collection_parameter :item

  def initialize(item: nil)
    @item = item
  end
end
```

Trying to render this component as a collection will raise a `MissingCollectionArgumentError`.

Comment (rainerborene) [reply lib/view_component/base.rb:652]: @camertron Got it. It's ready to merge now.

**Review by camertron — APPROVED** [review]

**Pull request merged**
